tools/CodeGen/StronglyTypedOverloadsGenerator.cs

```diff
@@ -0,0 +1,101 @@
+            //------------------------------------------------------------------------------
+            // <auto-generated>
+            //     This code was generated by {this.GetType().FullName}.
+            //
+            //     Changes to this file will be lost when the project is rebuilt.
+            // </auto-generated>
+            //------------------------------------------------------------------------------
+            {this.GenerateSource()}
+
+            """;
+        context.AddSource(
+            this.GetType().Name.Replace("Generator", ".g.cs"),
+            source);
+    }
+
+    public void Initialize(GeneratorInitializationContext context)
+    {
+        // No initialization required
+    }
+
+    protected static string OverloadMethod(Func<int, string> generator) =>
+        Escalate(MaximumTypeParameterCount, "\n\n", generator);
+
+    protected static string TypeParamDocsSource(int typeParametersCount) =>
+        Escalate(
+            typeParametersCount,
+            "\n",
+            i => $"""/// <typeparam name="T{i}">The type of the {Ordinals[i]} argument of the faked method call.</typeparam>""");
+
+    protected static string TypeParamList(int typeParametersCount) =>
+        Escalate(typeParametersCount, ", ", TypeParamName);
+
+    /// <summary>
+    /// Generates a string by invoking the <paramref name="generator" /> function for each number
+    /// from 1 to <paramref name="maximum" /> and joining the results
+    /// with <paramref name="separator"/>.
+    /// </summary>
+    /// <param name="maximum">The maximum number to generate strings for.</param>
+    /// <param name="separator">The separator to use between generated strings.</param>
+    /// <param name="generator">The function to generate strings.</param>
+    /// <returns>The generated string.</returns>
+    protected static string Escalate(int maximum, string separator, Func<int, string> generator) =>
+        string.Join(separator, Enumerable.Range(1, maximum).Select(generator));
+
+    protected static string Indent(int depth, string text)
+    {
+        var prefix = new string(' ', 4 * depth);
+        return string.Join(
+            "\r\n",
+            text.Split(["\r\n", "\n"], StringSplitOptions.None)
+                .Select(line => line.Length == 0 ? line : prefix + line));
+    }
+
+    protected static string ArgumentsFromCall(int typeParametersCount)
+    {
+        return Escalate(typeParametersCount, ", ", i => $"call.GetArgument<T{i}>({i - 1})!");
+    }
+
+    protected static bool IsTypeInTargetAssembly(GeneratorExecutionContext context, string fullyQualifiedMetadataName)
+    {
+        var type = context.Compilation.GetTypeByMetadataName(fullyQualifiedMetadataName);
+        return SymbolEqualityComparer.Default.Equals(context.Compilation.Assembly, type?.ContainingAssembly);
+    }
+
+    protected abstract string GenerateSource();
+
+    protected abstract bool IsApplicable(GeneratorExecutionContext context);

```

Comment:
An alternative could be to split the ValueTask generators into their own source generator and use the original generator only from FakeItEasy. The ValueTask generator would be used from both FIE and FIE.Extensions.ValueTask and would still need to detect whether it should run (since it shouldn't when building FIE for < netstandard2.1).

Reply:
But the generators would need to share some code... I think it's not worth it. I like the approach you took!